src/Microsoft.TestPlatform.Client/TestPlatform.cs

```diff
@@ -303,21 +277,133 @@ private void AddExtensionAssembliesFromSource(IEnumerable<string> sources)
+                || (strategy & TestAdapterLoadingStrategy.ExtensionsDirectory) == TestAdapterLoadingStrategy.ExtensionsDirectory)
+            {
+                defaultExtensionPaths = fileHelper
+                    .EnumerateFiles(extensionsFolder, SearchOption.TopDirectoryOnly, ".dll", ".exe")
+                    .Union(defaultExtensionPaths);
+            }
+        }
+
+        TestPluginCache.Instance.DefaultExtensionPaths = defaultExtensionPaths.Distinct();
+    }
+
+    private static SearchOption GetSearchOption(TestAdapterLoadingStrategy strategy, SearchOption defaultStrategyOption) {
+        if (strategy == TestAdapterLoadingStrategy.Default) {
+            return defaultStrategyOption;
+        }
+
+        var searchOption = SearchOption.TopDirectoryOnly;
+        if ((strategy & TestAdapterLoadingStrategy.Recursive) == TestAdapterLoadingStrategy.Recursive)
+        {
+            searchOption = SearchOption.AllDirectories;
+        }
+
+        return searchOption;     
+    }
 
-            TestPluginCache.Instance.DefaultExtensionPaths = defaultExtensionPaths;
+    private static IEnumerable<string> ExpandTestAdapterPaths(string path, IFileHelper fileHelper, TestAdapterLoadingStrategy strategy)
+    {
+        var adapterPath = Path.GetFullPath(Environment.ExpandEnvironmentVariables(path));
+
+        // Default behavior is to only accept directories!
+        if (strategy == TestAdapterLoadingStrategy.Default)
+        {
+            return ExpandAdaptersWithDefaultStrategy(adapterPath, fileHelper);
+        }
+
+        var adapters = ExpandAdaptersWithExplicitStrategy(adapterPath, fileHelper, strategy);
+
+        return adapters.Distinct();
+    }
+
+    private static IEnumerable<string> ExpandAdaptersWithExplicitStrategy(string path, IFileHelper fileHelper, TestAdapterLoadingStrategy strategy)
+    {
+        if ((strategy & TestAdapterLoadingStrategy.Explicit) != TestAdapterLoadingStrategy.Explicit)
+        {
+            return Enumerable.Empty<string>();
         }
+
+        if (fileHelper.Exists(path))
+        {
+            return new[] { path };
+        }
+        else if (fileHelper.DirectoryExists(path))
+        {
+            var searchOption = GetSearchOption(strategy, SearchOption.TopDirectoryOnly);
+
+            var adapterPaths = fileHelper.EnumerateFiles(
+                path,
+                searchOption,
+                TestPlatformConstants.TestAdapterEndsWithPattern,
+                TestPlatformConstants.TestLoggerEndsWithPattern,
+                TestPlatformConstants.DataCollectorEndsWithPattern,
+                TestPlatformConstants.RunTimeEndsWithPattern);
+
+            return adapterPaths;
+        }
+
+        EqtTrace.Warning(string.Format(CultureInfo.InvariantCulture, "AdapterPath Not Found: {0}", path));
+        return Enumerable.Empty<string>();
+    }
+
+    private static IEnumerable<string> ExpandAdaptersWithDefaultStrategy(string path, IFileHelper fileHelper)
```

Comment:
Again, feel free to ignore but I would rename `path` into `directoryPath` (or `folderPath`).